rust/wal3/src/lib.rs

```diff
@@ -0,0 +1,667 @@
+
+impl StreamID {
+    pub const INTERNAL: StreamID =
+        StreamID(Uuid::from_u128(0x610ba100d0600f00d00c0ffeec0ffee0_u128));
+}
+
+//////////////////////////////////////////// CursorName ////////////////////////////////////////////
+
+pub struct CursorName(pub String);
+
+////////////////////////////////////////////// Message /////////////////////////////////////////////
+
+#[derive(Clone, Debug)]
+pub enum Message {
+    Control(Vec<u8>),
+    Payload(Vec<u8>),
+}
+
+impl Message {
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn len(&self) -> usize {
+        match self {
+            Message::Control(v) => v.len(),
+            Message::Payload(v) => v.len(),
+        }
+    }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        match self {
+            Message::Control(v) => v.as_slice(),
+            Message::Payload(v) => v.as_slice(),
+        }
+    }
+}
+
+//////////////////////////////////////////// LogPosition ///////////////////////////////////////////
+
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    Default,
+    PartialEq,
+    Eq,
+    Ord,
+    PartialOrd,
+    serde::Deserialize,
+    serde::Serialize,
+)]
+pub struct LogPosition(pub u64);
+
+impl std::ops::Add<LogPosition> for usize {
+    type Output = LogPosition;
+
+    fn add(self, rhs: LogPosition) -> Self::Output {
+        LogPosition((self as u64).wrapping_add(rhs.0))
+    }
+}
+
+impl std::ops::Add<usize> for LogPosition {
+    type Output = LogPosition;
+
+    fn add(self, rhs: usize) -> Self::Output {
+        LogPosition(self.0.wrapping_add(rhs as u64))
+    }
+}
+
+impl std::ops::AddAssign<usize> for LogPosition {
+    fn add_assign(&mut self, rhs: usize) {
+        self.0 = self.0.wrapping_add(rhs as u64);
+    }
+}
+
+////////////////////////////////////////// ThrottleOptions /////////////////////////////////////////
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, arrrg_derive::CommandLine)]
+pub struct ThrottleOptions {
+    #[arrrg(optional, "Maximum batch size in bytes.")]
+    pub batch_size: usize,
+    #[arrrg(optional, "Minimum batch interval if not full in us. ")]
+    pub batch_interval: usize,
+    #[arrrg(optional, "Maximum throughput in ops/s.")]
+    pub throughput: usize,
+    #[arrrg(optional, "Maximum headroom in ops/s.")]
+    pub headroom: usize,
+    #[arrrg(optional, "Maximum number of outstanding ops.")]
+    pub outstanding: usize,
+}
+
+impl Default for ThrottleOptions {
+    fn default() -> Self {
+        ThrottleOptions {
+            // Batch for at least 20ms.
+            batch_interval: 20_000,
+            // Set a batch size of 8MB.
+            batch_size: 8 * 1_000_000,
+            // Set a throughput that's approximately 5/7th the throughput of the throughput S3
+            // allows.  If we hit throttle errors at this throughput we have a case for support.
+            throughput: 2_000,
+            // How much headroom we have for retries.
+            headroom: 1_500,
```

Comment:
I'm wondering if `throughput + headroom = <s3_throughput>`?

Reply:
That's how I picked these constants.